nixos/modules/services/web-apps/galene.nix

```diff
@@ -0,0 +1,130 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.galene;
+  defaultRecordingsRootPath = "${cfg.stateDir}/recordings";
+  defaultGroupsRootPath = "${cfg.stateDir}/groups";
+  defaultDataRootPath = "${cfg.stateDir}/data";
+in
+{
+  options = {
+    services.galene = {
+      enable = mkOption {
+        default = false;
+        type = types.bool;
+        description = "Enable Galene Service.";
+      };
+
+      stateDir = mkOption {
+        default = "/var/lib/galene";
+        type = types.str;
+        description = "Galene state directory.";
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "galene";
+        description = "User account under which galene runs.";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "galene";
+        description = "Group under which galene runs.";
+      };
+
+      httpAddress = mkOption {
+        type = types.str;
+        default = "";
+        description = "HTTP listen address for galene.";
+      };
+
+      httpPort = mkOption {
+        type = types.int;
+        default = 8443;
+        description = "HTTP listen port.";
+      };
+
+      staticRootPath = mkOption {
+        type = types.str;
+        default = "${pkgs.galene.static}/static";
+        example = "/var/lib/galene/static";
+        description = "Web server directory.";
+      };
+
+      recordingsRootPath = mkOption {
+        type = types.str;
+        default = defaultRecordingsRootPath;
+        example = "/var/lib/galene/recordings";
+        description = "Recordings directory.";
+      };
+
+      dataRootPath = mkOption {
+        type = types.str;
+        default = defaultDataRootPath;
+        example = "/var/lib/galene/data";
+        description = "Data directory.";
+      };
+
+      groupsRootPath = mkOption {
+        type = types.str;
+        default = defaultGroupsRootPath;
+        example = "/var/lib/galene/groups";
+        description = "Web server directory.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.galene = {
+      description = "galene";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      path = [ pkgs.galene ];
+
+      preStart =
+        ''
+          mkdir -m 0700 -p ${cfg.stateDir}
```

Comment:
Use `systemd-tmpfiles` rather than this approach

Reply:
`systemd-tmpfiles` is a much better solution than `mkdir`. With what you have here `mkdir` will fail the first time around because there are no permissions for the `galene` user to create `/var/lib/galene`.

Even better than `systemd-tmpfiles`, though, would be to utilize `StateDirectory`, like so:
```
options = {
  services.galene = {
    stateDir = mkOption {
      default = "/var/lib/galene";
      type = types.str;
      description = ''
        The directory where Galene stores its internal state. If left as the default
        value this directory will automatically be created before the Galene server
        starts, otherwise the sysadmin is responsible for ensuring the directory
        exists with appropriate ownership and permissions.
      '';
    };
  };
};
config = mkIf cfg.enable {
  systemd.services.galene = {
    ...
    serviceConfig = [
      { Type = "simple";
        User = cfg.user;
        Group = cfg.group;
        ...
      }
      (mkIf (cfg.stateDir == "/var/lib/galene") { StateDirectory = "galene"; })
    ];
  };
}
```

Reply:
I feel meh with this solution, at least, it would be great to just look if `cfg.stateDir` starts with `/var/lib` and use `StateDirectory` in this case, otherwise, just letting the sysadmin create the folder themselves.

Reply:
While it is just a suggestion I feel it flows nicely with existing NixOS patterns. Note how users and groups are usually only provisioned if the default value is provided.

Take your pick, though.

Reply:
I went with your solution, @aanderse . I still use tmpfiles in the meantime for subdirectories, and for the management of the private key and certificate files: Galene supposes the private key and the cert have specific name and are in a specific directory, which I would like to be configurable too. So, I tried something with tmpfiles, I'm waiting for your feedbacks on this, this is probably a bit weird.